Add tests for cloned kernel via clCloneKernel

MichaelRizkalla-arm opened the pull request.

MichaelRizkalla-arm committed:
Add test coverage for `clCloneKernel` with args

Use `clSetKernelArg` to set args after kernel is cloned. Enqueue and
read the buffer to validate.
The test uses `buf_write_kernel` kernel program with 2 arguments.

Signed-off-by: Michael Rizkalla <michael.rizkalla@arm.com>
Co-authored-by: Vikas Katariya <vikas.katariya@arm.com>
Co-authored-by: Ahmed Hesham <ahmed.hesham@arm.com>

MichaelRizkalla-arm committed:
Add test coverage for `clCloneKernel` with Execinfo

Use `clSetKernelExecInfo` after kernel is cloned and read the buffer
to validate.

The test uses the `set_kernel_exec_info_kernel` kernel program
with 2 arguments.

Signed-off-by: Michael Rizkalla <michael.rizkalla@arm.com>
Co-authored-by: Vikas Katariya <vikas.katariya@arm.com>
Co-authored-by: Ahmed Hesham <ahmed.hesham@arm.com>

MichaelRizkalla-arm committed:
Add test covergae for `clCloneKernel` with no args

Clone a kernel with no args and enqueue.
The test uses `test_kernel_empty` kernel program with no arguments.

Signed-off-by: Michael Rizkalla <michael.rizkalla@arm.com>
Co-authored-by: Vikas Katariya <vikas.katariya@arm.com>
Co-authored-by: Ahmed Hesham <ahmed.hesham@arm.com>

MichaelRizkalla-arm committed:
Add test coverge for `clCloneKernel` with SVM pointer

Use `clSetKernelArgSVMPointer` to set args after kernel is cloned.
Enqueue and read the buffer to validate.
The test uses `buf_write_kernel` kernel program with 2 arguments.

Signed-off-by: Michael Rizkalla <michael.rizkalla@arm.com>
Co-authored-by: Vikas Katariya <vikas.katariya@arm.com>
Co-authored-by: Ahmed Hesham <ahmed.hesham@arm.com>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/KhronosGroup/OpenCL-CTS?pullRequest=2577) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/KhronosGroup/OpenCL-CTS?pullRequest=2577) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/KhronosGroup/OpenCL-CTS?pullRequest=2577) it.</sub>

EwanC commented on test_conformance/api/test_clone_kernel.cpp:
```diff
@@ -346,56 +463,462 @@ REGISTER_TEST_VERSION(clone_kernel, Version(2, 1))
+}
+
+int test_exec_enqueue_helper(cl_context context, cl_command_queue queue,
+                             BufPtr* pBuf, cl_int* svmPtr_Kernel,
```
`pBuf` looks unused int this function

MichaelRizkalla-arm commented on test_conformance/api/test_clone_kernel.cpp:
```diff
@@ -346,56 +463,462 @@ REGISTER_TEST_VERSION(clone_kernel, Version(2, 1))
+                                      &srcKernel, &intargs[0])
+            != TEST_PASS)
+        {
+            test_fail("test_svm_exec_info_helper failed for srcKernel.\n");
```
I've added a struct that will clean-up the allocations.

EwanC commented on test_conformance/api/test_clone_kernel.cpp:
```diff
@@ -324,78 +453,471 @@ REGISTER_TEST_VERSION(clone_kernel, Version(2, 1))
+
+    error = clFinish(queue);
+    test_error(error, "clFinish failed");
+
```
Do we need this `clFinish` call if we've already got one a few lines above before setting the arguments again?

MichaelRizkalla-arm replied:
It's redundant, I've removed it.

EwanC reviewed (dismissed):
I haven't had a chance to run this code, but LGTM with some minor comments

bashbaug commented on test_conformance/api/test_clone_kernel.cpp:
```diff
@@ -324,78 +453,471 @@ REGISTER_TEST_VERSION(clone_kernel, Version(2, 1))
+    error = clSetKernelArg(srcKernel, 0, sizeof(cl_int), &value);
+    test_error(error, "clSetKernelArg failed");
+    error = clSetKernelArgSVMPointer(srcKernel, 1, pBuf);
+    test_error(error, "clSetKernelArgSVMPointer failed");
```
This is called `test_svm_exec_info_helper` but I don't see any place where `clSetKernelExecInfo` is called.

Since we're passing `svmPtr_Kernel` indirectly, I think we need something like:


```suggestion
    error = clSetKernelArgSVMPointer(srcKernel, 1, pBuf);
    test_error(error, "clSetKernelArgSVMPointer failed");

    error = clSetKernelExecInfo(srcKernel, CL_KERNEL_EXEC_INFO_SVM_PTRS,
                                sizeof(svmPtr_Kernel), &svmPtr_Kernel);
    test_error(error, "clSetKernelExecInfo failed");
```

MichaelRizkalla-arm replied:
Done.

bashbaug commented:
Removing "focused review" until the comments above are addressed.

MichaelRizkalla-arm committed:
Add equality operators to clSVMWrapper

Add equality operators to `clSVMWrapper` to compare with `clSVMWrapper`
object or a `nullptr`.

Signed-off-by: Michael Rizkalla <michael.rizkalla@arm.com>

MichaelRizkalla-arm commented on test_conformance/api/test_clone_kernel.cpp:
```diff
@@ -324,78 +453,471 @@ REGISTER_TEST_VERSION(clone_kernel, Version(2, 1))
+    test_error(error, "clSetKernelArg failed");
+
+    error = clFinish(queue);
+    test_error(error, "clFinish failed");
```
I removed it.

bashbaug commented:
Latest version LGTM - thanks!  @lakshmih is this one you'd like to review?

bashbaug commented:
Discussed in the May 26th teleconference.  OK to merge offline with one additional review (@lakshmih).

rjodinchr commented:
This commit broke some tests when running with `clvk` which is an OpenCL3.0 implementation with a OpenCL-C1.2 compiler.
Here is the fix for it: https://github.com/KhronosGroup/OpenCL-CTS/pull/2712
Please take a look, thank you
